Kernel/FileSystem/Ext2FS/Inode.cpp

```diff
@@ -895,18 +933,30 @@ ErrorOr<void> Ext2FSInode::update_timestamps(Optional<UnixDateTime> atime, Optio
     MutexLocker locker(m_inode_lock);
     if (fs().is_readonly())
         return EROFS;
-    if (atime.value_or({}).to_timespec().tv_sec > NumericLimits<i32>::max())
+    if (atime.value_or({}).to_timespec().tv_sec > NumericLimits<i32>::max() && !is_within_inode_bounds(bit_cast<FlatPtr>(&m_raw_inode), bit_cast<FlatPtr>(&m_raw_inode.i_atime_extra), sizeof(m_raw_inode.i_atime_extra)))
         return EINVAL;
-    if (ctime.value_or({}).to_timespec().tv_sec > NumericLimits<i32>::max())
+    if (ctime.value_or({}).to_timespec().tv_sec > NumericLimits<i32>::max() && !is_within_inode_bounds(bit_cast<FlatPtr>(&m_raw_inode), bit_cast<FlatPtr>(&m_raw_inode.i_ctime_extra), sizeof(m_raw_inode.i_ctime_extra)))
         return EINVAL;
-    if (mtime.value_or({}).to_timespec().tv_sec > NumericLimits<i32>::max())
+    if (mtime.value_or({}).to_timespec().tv_sec > NumericLimits<i32>::max() && !is_within_inode_bounds(bit_cast<FlatPtr>(&m_raw_inode), bit_cast<FlatPtr>(&m_raw_inode.i_mtime_extra), sizeof(m_raw_inode.i_mtime_extra)))
         return EINVAL;
-    if (atime.has_value())
-        m_raw_inode.i_atime = atime.value().to_timespec().tv_sec;
-    if (ctime.has_value())
-        m_raw_inode.i_ctime = ctime.value().to_timespec().tv_sec;
-    if (mtime.has_value())
-        m_raw_inode.i_mtime = mtime.value().to_timespec().tv_sec;
+    if (atime.has_value()) {
+        time_t seconds = atime.value().to_timespec().tv_sec;
+        u32 nanoseconds = atime.value().to_timespec().tv_nsec;
+        m_raw_inode.i_atime = static_cast<u32>(seconds);
+        m_raw_inode.i_atime_extra = encode_time_to_extra(seconds, nanoseconds);
+    }
+    if (ctime.has_value()) {
```

Comment:
This feels pretty repetitive. Could be shorter with a local lambda for the common code. (Further up too, to a lesser degree.)